feat: Add Human readable descriptions

usame-algan opened the pull request:
## What it solves

First part of the EPIC #2011 

Resolves #2448 

Depends on https://github.com/safe-global/safe-client-gateway-nest/pull/631

## How this PR fixes it

- Displays human-readable descriptions for nonce replacement and notifications

## ToDos

- [x] Update gateway-typescript-sdk type for `TransactionInfo`
- [x] Check remaining `dispatch` functions if they need a `txDescription`


## How to test it

1. Queue a transaction (transfer, settings change, custom transaction)
2. Create a new transaction
3. Open the nonce replacement form
4. Observe human-readable descriptions
5. If there is no human-readable description observe the same message as before
6. Sign or execute a transaction
7. Observe a human-readable description inside notifications and the notification center
8. If there is no human-readable description observe a generic fallback saying `Transaction` as the Notification title

## Screenshots
<img width="366" alt="Screenshot 2023-08-11 at 15 41 52" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/ea503859-61b4-434d-9083-74976c82fca6">
<img width="471" alt="Screenshot 2023-08-11 at 17 22 29" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/4a0510d2-df41-40dc-a8a2-e50a4716bff8">

<img width="380" alt="Screenshot 2023-08-11 at 13 18 53" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/cf50ab45-b31f-4688-a459-eb98b59b363d">


## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

iamacook commented on src/config/constants.ts:
```diff
@@ -3,9 +3,8 @@ import chains from './chains'
-export const GATEWAY_URL_PRODUCTION =
-  process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'https://safe-client.safe.global'
-export const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'https://safe-client.staging.5afe.dev'
+export const GATEWAY_URL_PRODUCTION = process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'http://localhost:3030'
```
Not review related: did you change the default gateway port? It opens on `3000` for me.

usame-algan replied:
Yes, I used a custom port 3030 with `APPLICATION_PORT=3030 yarn start:dev`

iamacook commented on .vscode/settings.json:
```diff
@@ -1,4 +0,0 @@
-{
```
Perhaps we should add this to the `.gitignore`?

katspaugh commented on src/components/tx/SignOrExecuteForm/hooks.ts:
```diff
@@ -18,10 +18,11 @@ import type { OnboardAPI } from '@web3-onboard/core'
 type TxActions = {
-  addToBatch: (safeTx?: SafeTransaction, origin?: string) => Promise<string>
-  signTx: (safeTx?: SafeTransaction, txId?: string, origin?: string) => Promise<string>
+  addToBatch: (safeTx?: SafeTransaction, origin?: string, humanDescription?: string) => Promise<string>
+  signTx: (safeTx?: SafeTransaction, txId?: string, origin?: string, humanDescription?: string) => Promise<string>
```
Passing `humanDescription` to these two methods seems oddly specific. I would suggest passing the whole `txDetails`.

usame-algan replied:
I've changed the signature to accept a transaction object instead.

katspaugh commented on src/hooks/useTxNotifications.ts:
```diff
@@ -79,9 +76,11 @@ const useTxNotifications = (): void => {
 
         dispatch(
           showNotification({
+            title: humanDescription,
```
What will it look like w/o a title?

usame-algan replied:
I went through each tx event that we show a notification for and made sure to add a humanDescription there, either from the CGW response or an appropriate fallback i.e. `Transaction #<nonce>` or `Module transaction`

katspaugh commented on src/services/tx/tx-sender/dispatch.ts:
```diff
@@ -50,7 +50,10 @@ export const dispatchTxProposal = async ({
   } catch (error) {
     if (txId) {
-      txDispatch(TxEvent.SIGNATURE_PROPOSE_FAILED, { txId, error: asError(error) })
+      txDispatch(TxEvent.SIGNATURE_PROPOSE_FAILED, {
+        txId,
+        error: asError(error),
+      })
```
Unnecessary formatting
```suggestion
      txDispatch(TxEvent.SIGNATURE_PROPOSE_FAILED, { txId, error: asError(error) })
```

iamacook commented on src/hooks/useTxNotifications.ts:
```diff
@@ -16,23 +16,20 @@ import useSafeAddress from './useSafeAddress'
-  [TxEvent.REVERTED]: 'Transaction reverted. Please check your gas settings.',
-  [TxEvent.SUCCESS]: 'Your transaction was successfully executed.',
-  [TxEvent.FAILED]: 'Your transaction was unsuccessful.',
+  [TxEvent.SIGN_FAILED]: 'Failed to sign. Please try again.',
+  [TxEvent.PROPOSED]: 'Successfully added to queue.',
+  [TxEvent.PROPOSE_FAILED]: 'Failed to add to queue. Please try again.',
+  [TxEvent.SIGNATURE_PROPOSED]: 'Successfully signed.',
+  [TxEvent.SIGNATURE_PROPOSE_FAILED]: 'Failed to send signature. Please try again.',
+  [TxEvent.EXECUTING]: 'Confirm the execution in your wallet.',
+  [TxEvent.PROCESSING]: 'Validating...',
+  [TxEvent.PROCESSING_MODULE]: 'Validating module interaction...',
+  [TxEvent.ONCHAIN_SIGNATURE_REQUESTED]: 'Confirm on-chain signature in your wallet.',
+  [TxEvent.ONCHAIN_SIGNATURE_SUCCESS]: 'On-chain signature request confirmed.',
+  [TxEvent.PROCESSED]: 'Successfully validated. Indexing...',
+  [TxEvent.REVERTED]: 'Reverted. Please check your gas settings.',
+  [TxEvent.SUCCESS]: 'Successfully executed.',
+  [TxEvent.FAILED]: 'Failed.',
```
Why are these changes necessary? If there's no `humanDescription`, we now lose some context.

katspaugh replied:
`humanDescription` will be always there unless the feature is disabled for technical reasons.

iamacook replied:
If the signature hash does not match on the gateway the `humanDescription` will be `null`. You can test this with spending limits atm:

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/4fd7a4eb-18bd-4f24-985b-0c6965ebc959)

usame-algan replied:
Good point, I think we should add a fallback for the title. Maybe we can include the nonce, something like `Transaction #113`

usame-algan commented on src/hooks/useTransactionType.ts:
```diff
@@ -80,13 +84,13 @@ export const getTransactionType = (tx: TransactionSummary, addressBook: AddressB
       return {
         icon: toAddress?.logoUri || '/images/transactions/custom.svg',
-        text: addressBookName || toAddress?.name || 'Contract interaction',
+        text: tx.txInfo.humanDescription || addressBookName || toAddress?.name || 'Contract interaction',
```
I am reverting this file since this is something we wan't to include in v2 as discussed.

iamacook commented on src/components/tx/SignOrExecuteForm/hooks.ts:
```diff
@@ -83,50 +85,55 @@ export const useTxActions = (): TxActions => {
         const id = txId || (await proposeTx(wallet.address, safeTx, txId, origin)).txId
-        await dispatchOnChainSigning(safeTx, id, onboard, chainId)
+        await dispatchOnChainSigning(safeTx, id, onboard, chainId, humanDescription)
         return id
       }
 
       // Otherwise, sign off-chain
-      const signedTx = await dispatchTxSigning(safeTx, version, onboard, chainId, txId)
+      const signedTx = await dispatchTxSigning(safeTx, version, onboard, chainId, txId, humanDescription)
```
Should we also pass the whole transaction object here instead of `humanDescription`?

usame-algan replied:
We have all the info to build a humanDescription outside of the dispatch functions so I would suggest we keep all the logic there and only pass the string to keep those functions as pure as possible. I would move the fallbacks outside as well.

iamacook commented on src/config/constants.ts:
```diff
@@ -3,9 +3,8 @@ import chains from './chains'
-export const GATEWAY_URL_PRODUCTION =
-  process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'https://safe-client.safe.global'
-export const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'https://safe-client.staging.5afe.dev'
+export const GATEWAY_URL_PRODUCTION = process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'http://localhost:3030'
+export const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'http://localhost:3030'
```
I'm unable to test on production with these fallbacks still set.

iamacook commented on src/services/tx/tx-sender/dispatch.ts:
```diff
@@ -191,37 +200,39 @@ export const dispatchBatchExecution = async (
 
-    txs.forEach(({ txId }) => {
-      txDispatch(TxEvent.EXECUTING, { txId, groupKey })
+    txs.forEach(({ txId, txInfo: { humanDescription } }) => {
```
Do we need a fallback for this dispatcher?

usame-algan replied:
I've changed it to a static description `Batch transaction`. Imo it is excessive to show a notification for every transaction but this is something we could also change later.

francovenica commented:
This message is the same as the one in stg currently. in this case I tried to execute a batch tx (2 fully signed tx in queue, and using the Execute batch button), but these tx are incompatible so is expected for the batch to fail:

Idk if something more user friendly can be set, or if the information in the tx means something for the user.
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/ad500808-1a5d-45bb-ade9-76d4d19eb30a)

francovenica commented:
just "send"? in the description it shows that it should show what token you are sending, how much and to who.
In this case I'm testing in Goerli, sending GoerliETH
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/49367781-3413-46d9-80aa-764dbbdca7f0)

usame-algan commented on src/services/tx/tx-sender/dispatch.ts:
```diff
@@ -49,10 +49,11 @@ export const dispatchTxProposal = async ({
   try {
     proposedTx = await proposeTx(chainId, safeAddress, sender, safeTx, safeTxHash, origin)
   } catch (error) {
+    const humanDescription = `Transaction #${safeTx.data.nonce}`
```
I've moved this outside of this dispatch call so we have the human description + fallback in one place.

usame-algan commented on src/services/tx/tx-sender/dispatch.ts:
```diff
@@ -316,6 +334,7 @@ export const dispatchTxRelay = async (
   safe: SafeInfo,
   txId: string,
   gasLimit?: string | number,
+  humanDescription?: string,
```
I've added the same fallback as for the execution

usame-algan commented on src/store/txHistorySlice.ts:
```diff
@@ -28,7 +28,11 @@ export const txHistoryListener = (listenerMiddleware: typeof listenerMiddlewareI
+          txDispatch(TxEvent.SUCCESS, {
+            txId,
+            groupKey: pendingTxs[txId].groupKey,
+            humanDescription: result.transaction.txInfo?.humanDescription,
```
Thanks, I've added a fallback.

usame-algan commented:
@francovenica thanks for testing! I've fixed the "Send" issue. This was a regression, nice find! For Batch Executions I now added a static description so it should always say "Batch transaction" as the title.

iamacook commented:
> @iamacook @katspaugh what do you think about removing all the custom fallbacks and only having one fallback to "Transaction" inside `useTxNotifications`. We would lose some information like the nonce but I don't think its too critical since thats how it was before.

Having a SSOT makes the most sense.

francovenica commented:
I tried the tx that Usame shared with me.
I was able to see most of them

I wasn't able to trigger "ONCHAIN_SIGNATURE_REQUESTED" and "ONCHAIN_SIGNATURE_SUCCESS". I'm gonna throw that, these probably were used back when every signature was on chain, even the ones that proposed a tx without execution.

The issue I reported about the nonce replacement message now show the tx name correctly

check the size of the nonce input tx display for mobile, looks fine as well:
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/2387a342-f5d3-4b75-9bf5-41e7317a1f67)

Looks good to me

usame-algan commented:
> Found an issue a last minute. a error message that show a "GS013" error

Nice find! There was an issue with setting the fallback title. It should be fixed now and display `Transaction` at the top of the notification.

francovenica commented:
Yes, it has it now:
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/b8ef28e0-f5c0-4e13-a3d3-9dcf8e712b86)

francovenica commented:
Ok, looks good to me. I can pass this to done now.
